Clever/workflow-manager
# support step functions

## POC: workflows via AWS Step Functions

This PR creates a new implementation of the `WorkflowManager` interface that uses AWS Step Functions.

Some advantages of going the way of Step Functions:
1. It supports the entirety of the states language spec (timeouts, retries, input/output parsing, etc.)
2. We can continue to run our code in our us-west-1 ECS cluster, no network tunneling or additional log routing infra required.
3. We get a pretty good [GUI](https://us-west-2.console.aws.amazon.com/states/home?region=us-west-2#/) out of the box
4. We can swap `gearcmd` for `sfncli` and run/deploy workers the same way we do today
5. We get [CW metrics](http://docs.aws.amazon.com/step-functions/latest/dg/procedure-cw-metrics.html) on workflows, which we can alert on.

## What the PR does to the code

1. Implements a new `WorkflowManager` that uses Step Functions (`SFNWorkflowManager`).
2. Implements a new `WorkflowManager` that multiplexes across other implementations (`MultiWorkflowManager`).
3. Adds a `Manager` field to workflow definitions that is used to determine which `WorkflowManager` implementation to use.
4. Adds a `Queue` and `Namespace` field to the `Workflow` object. This was needed to identify executions in the AWS SFN API.

## Current state of this PR: I have it working locally, in theory it should work once deployed as well. How to use it locally:

1. Start this up, pointing at dev dynamo tables so we can create workflow definitions:

  ```
ark start -l -e clever-dev -s all-dev-dynamo-tables
  ```

2. Create a workflow definition with manager set to "step-functions":

  ```
curl -XPOST localhost:8080/workflow-definitions -d '{
    "name": "echo-test",
    "startAt": "echo",
    "states": [
        {
            "name": "echo",
            "resource": "echo",
            "type": "Task",
            "end": true
        }
    ],
    "manager": "step-functions"
}'
  ```

3. Start a worker locally:

  ```
sfncli -cmd echo -name raf-echo -region us-west-2
  ```

4. Begin a workflow:

  ```
curl -XPOST localhost:8080/workflows -d '{
  "workflowDefinition": { "name": "echo-test", "version": -1 },
  "data": ["hello", "world"],
  "namespace": "raf",
  "queue": "default"
}'
  ```

4. See the workflow succeed in the console:

![image](https://user-images.githubusercontent.com/72655/30530117-d9fb4774-9bf8-11e7-972f-2a4cbda572a2.png)

🎉
